lightning/src/ln/channelmanager.rs

```diff
@@ -6219,6 +6225,269 @@ where
+									}],
+									incoming_trampoline_shared_secret,
+									hops: Vec::new(),
+								};
+
+								let mut push_trampoline_forwarding_failure = |msg: String, htlc_source: HTLCSource, forward_scid: Option<u64>, reason: LocalHTLCFailureReason, err_data: Vec<u8>| {
+									let logger = WithContext::from(&self.logger, Some(next_node_id), Some(prev_channel_id), Some(payment_hash));
+									log_info!(logger, "Failed to forward incoming Trampoline HTLC: {}", msg);
+
+									failed_forwards.push((htlc_source, payment_hash,
+										HTLCFailReason::reason(reason, err_data),
+										HTLCDestination::FailedTrampolineForward { requested_next_node_id: next_node_id, forward_scid }
+									));
+								};
+
+								let next_blinding_point = blinded.and_then(|b| {
+									b.next_blinding_override.or_else(|| {
+										let encrypted_tlvs_ss = self.node_signer.ecdh(
+											Recipient::Node, &b.inbound_blinding_point, None
+										).unwrap().secret_bytes();
+										onion_utils::next_hop_pubkey(
+											&self.secp_ctx, b.inbound_blinding_point, &encrypted_tlvs_ss
+										).ok()
+									})
+								});
+
+								let incoming_amount = match incoming_amt_msat {
+									Some(amount) => amount,
+									None => {
+										push_trampoline_forwarding_failure(format!("Missing incoming amount to calculate routing parameters to next Trampoline hop {next_node_id}"), htlc_source, None, LocalHTLCFailureReason::TemporaryTrampolineFailure, Vec::new());
+										continue;
+									}
+								};
+
+								let proportional_fee = self.default_configuration.channel_config.forwarding_fee_proportional_millionths as u64 * outgoing_amt_msat / 1_000_000;
+								let forwarding_fee = proportional_fee + self.default_configuration.channel_config.forwarding_fee_base_msat as u64;
+								let cltv_expiry_delta = incoming_cltv_expiry - outgoing_cltv_value;
+
+								let max_total_routing_fee_msat = match incoming_amount.checked_sub(forwarding_fee + outgoing_amt_msat) {
+									Some(amount) => amount,
+									None => {
+										let mut data = Vec::new();
+										self.default_configuration.channel_config.forwarding_fee_base_msat.write(&mut data);
+										self.default_configuration.channel_config.forwarding_fee_proportional_millionths.write(&mut data);
+										// todo: error handling
+										u16::try_from(cltv_expiry_delta).unwrap().write(&mut data);
+										push_trampoline_forwarding_failure(format!("Insufficient fee to forward to the next Trampoline hop {next_node_id}"), htlc_source, None, LocalHTLCFailureReason::TrampolineFeeOrExpiryInsufficient, Vec::new());
+										continue;
+									}
+								};
+
+								let usable_channels: Vec<ChannelDetails> = self.list_usable_channels();
+
+								// assume any Trampoline node supports MPP
+								let mut recipient_features = Bolt11InvoiceFeatures::empty();
+								recipient_features.set_basic_mpp_optional();
+
+								let route = match self.router.find_route(
```

Comment:
One question that came to my mind is to what extent pathfinding here is a DoS vector. It's probably quite easy to keep trampoline nodes busy searching?